File: redwood-yoga-java/src/main/java/io/github/orioncraftmc/meditate/internal/detail/GlobalMembers.java
package io.github.orioncraftmc.meditate.internal.detail;

import static io.github.orioncraftmc.meditate.internal.GlobalMembers.YGConfigGetDefault;
import io.github.orioncraftmc.meditate.internal.YGConfig;
import io.github.orioncraftmc.meditate.internal.YGNode;
import io.github.orioncraftmc.meditate.internal.enums.YGLogLevel;
import java.util.Arrays;
import java.util.Map;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class GlobalMembers {
    public static void vlog(@Nullable YGConfig config, YGNode node, YGLogLevel level, Object context, String format, Object... args) {
        @Nullable YGConfig logConfig = config != null ? config : YGConfigGetDefault();
        logConfig.log(logConfig, node, level, context, format, args);
    }

    public static @NotNull Integer log2ceilFn(Integer n) {
        return n < 1 ? 0 : (1 + log2ceilFn(n / 2));
    }

    public static int mask(Integer bitWidth, Integer index) {
        return ((1 << bitWidth) - 1) << index;
    }

    public static <E extends Enum<E>> int bitWidthFn(@NotNull Class<E> e) {
        return (log2ceilFn(e.getEnumConstants().length - 1));
    }

    public static <E extends Enum<E>> E getEnumData(@NotNull Class<E> e, Map<Object, Object> flags, Integer index) {
        return (E) flags.getOrDefault(new StyleEnumFlagsKey(e, index),
                Arrays.stream(e.getEnumConstants()).findFirst().get());
    }


    public static <E extends Enum<E>> int setEnumData(@NotNull Class<E> e, Map<Object, Object> flags, int index, @NotNull E newValue) {
        flags.put(new StyleEnumFlagsKey(e, index), newValue);
        return 0;
    }

    public static boolean getBooleanData(Map<Object, Object> flags, Integer index) {
        return (boolean) flags.getOrDefault(index, false);
    }

    public static int setBooleanData(Map<Object, Object> flags, int index, boolean value) {
        flags.put(index, value);
        return 0;
    }

    private boolean notEqualsTo(@NotNull CompactValue a, @NotNull CompactValue b) {
        return !CompactValue.equalsTo(a, b);
    }
}

File: redwood-yoga/src/commonMain/kotlin/app/cash/redwood/yoga/detail/GlobalMembers.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
package app.cash.redwood.yoga.detail

import app.cash.redwood.yoga.GlobalMembers
import app.cash.redwood.yoga.YGConfig
import app.cash.redwood.yoga.YGNode
import app.cash.redwood.yoga.enums.YGLogLevel
import kotlin.reflect.KClass

object GlobalMembers {
  fun vlog(
    config: YGConfig?,
    node: YGNode?,
    level: YGLogLevel,
    context: Any?,
    format: String,
    vararg args: Any?,
  ) {
    val logConfig = config ?: GlobalMembers.YGConfigGetDefault()
    logConfig.log(logConfig, node, level, context, format, *args)
  }

  fun log2ceilFn(n: Int): Int {
    return if (n < 1) 0 else 1 + log2ceilFn(n / 2)
  }

  fun mask(bitWidth: Int, index: Int): Int {
    return (1 shl bitWidth) - 1 shl index
  }

  inline fun <reified E : Enum<E>> bitWidthFn(): Int {
    return bitWidthFn(enumValues<E>())
  }

  fun <E : Enum<E>> bitWidthFn(values: Array<E>): Int {
    return log2ceilFn(values.size - 1)
  }

  inline fun <reified E : Enum<E>> getEnumData(
    e: KClass<E>,
    flags: Map<Any?, Any>,
    index: Int?,
  ): E {
    return getEnumData(e, enumValues(), flags, index)
  }

  fun <E : Enum<E>> getEnumData(
    e: KClass<E>,
    values: Array<E>,
    flags: Map<Any?, Any>,
    index: Int?,
  ): E {
    return (flags[StyleEnumFlagsKey(e, index!!)] ?: values.first()) as E
  }

  fun <E : Enum<E>> setEnumData(
    e: KClass<out E>,
    flags: MutableMap<Any?, Any>,
    index: Int,
    newValue: E,
  ): Int {
    flags[StyleEnumFlagsKey(e, index)] = newValue
    return 0
  }

  fun getBooleanData(flags: Map<Any?, Any>, index: Int?): Boolean {
    return (flags[index] ?: false) as Boolean
  }

  fun setBooleanData(flags: MutableMap<Any?, Any>, index: Int, value: Boolean): Int {
    flags[index] = value
    return 0
  }
}
